Return 404 on missing modules

mat128 opened the pull request.

Commit:
Create specific routes for every module

This will make Flask return a 404 when required and will solve issue #1

mbelang commented:
:+1:

ardeois commented:
You could have returned a 404 instead of having a real `favicon.ico` on an API

mat128 commented:
@ardeois Any clean way to handle this? Flask doesnt seem to help me except defining a specific route for it. In that case, I'd rather just not handle favicon and I shouldn't care about people querying the API with their browser

mat128 commented:
Did you look at the favicon?

fguillot commented:
@mat128 The problem is not only for `http://hostname/favicon.ico` but for any module that doesn't exists, example: `http://hostname/patate/` will also generate the same exception.

``` python
ERROR:ubersmith_remote_module_server.server:Exception on /patate/ [GET]
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/site-packages/flask/app.py", line 1817, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python3.5/site-packages/flask/app.py", line 1477, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python3.5/site-packages/flask/app.py", line 1381, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.5/site-packages/flask/_compat.py", line 33, in reraise
    raise value
  File "/usr/local/lib/python3.5/site-packages/flask/app.py", line 1475, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.5/site-packages/flask/app.py", line 1461, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/usr/local/lib/python3.5/site-packages/ubersmith_remote_module_server/api.py", line 30, in list_implemented_methods
    'implemented_methods': self.router.list_implemented_methods(module_name)
  File "/usr/local/lib/python3.5/site-packages/ubersmith_remote_module_server/router.py", line 35, in list_implemented_methods
    module = self.modules[module_name]
KeyError: 'patate'
```

You should check that the module name is present in `self.modules`

JoProvost commented:
Looks good to me! :+1:  :trollface:

lindycoder commented:
:-1: This is not an issue

This service is meant to be an RPC server, it responds to GET / and post / that's it, anything else should be completely discarded.

There's no reason to go onto that endpoint with a browser. so why do we need a favicon?

ardeois commented:
Right, like fguillot said, you can generate an exception like `NotFoundException` that extends `HTTPException`:

``` python
from werkzeug.exceptions import HTTPException

class NotFoundException(HTTPException):
    code = 404
    description = 'The requested resource could not be found.'
```

Flask by default will handle `HTTPException` and respond the proper HTTP code

mat128 commented:
Alright, I took your comments and here's the result:
- No specific favicon handling (I agree with you @lindycoder, this isnt a browser thing); however:
- Flask/Werkzeug will redirect anything that looks like a match that is missing the trailing slash to the proper route (and this behavior is almost impossible to disable (strict_slashes lets you choose between an external redirect (301) or an internal route redirect) so querying the favicon.ico will request the list of implemented methods on the "favicon.ico" module.
- Unknown modules correctly raise and the API layer catches that to return a werkzeug NotFound exception, which maps to a 404 (thanks @ardeois). This will implement the correct behavior, fixing the issue reported by @fguillot

lindycoder commented:
well you could have added routes as you serve them and let flask do the 404 itself.

So when you .serve("mymodule')
you could do .add_url("/{}".format(module_name))
instead of having a generic /<module_name> where you need to do shenanigans

mat128 commented:
Thanks to @lindycoder's help, I have been able to implement specific routes for every module and have avoided to need to manually manage 404s. The API is now responsible for modules and will pass them to the router along with necessary call information. The router will now only be responsible of the routing logic (yay!).

lindycoder commented on ubersmith_remote_module_server/api.py:
```diff
@@ -14,25 +14,30 @@
+        for module_name, module in modules.items():
+            list_endpoint = functools.partial(self.list_implemented_methods, module)
+            list_endpoint.__name__ = "list_" + module_name
+            app.add_url_rule('/{}/'.format(module_name), view_func=list_endpoint, methods=['GET'])
```
why is there still a trailing slash here ?

Commit:
Handle non-trailing-slash routes as well

The "official" spec doesn't specify how to behave on a multiple-module endpoint like this one. Follow the least surprise principle and make both work.

mat128 commented:
That should do it. Any reviews on that final v4 version? :)

fguillot commented:
:+1:

fguillot commented:
:+1:

lindycoder commented:
:+1:
